### PR TITLE
SG-584 The bucket listing breaks down after bucket creation and deletion.

### PR DESCRIPTION
**Pull request by dmitrizpanasas**
## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/panfs.go [hunk 1]
```diff
@@ -1480,7 +1480,7 @@ func (fs *PANFSObjects) DeleteObject(ctx context.Context, bucket, object string,
 		noLock := ""
 		err = fs.configAgent.DeleteObject(objectName, noLock)
 	} else {
-		err = fsDeleteFile(ctx, pathJoin(bucketDir), pathJoin(bucketDir, object))
```
Comment (lech-lorens-epam) [inline cmd/panfs.go]: Please, add a comment explaining why this change was necessary.

Comment (lech-lorens-epam) [reply cmd/panfs.go]: By changing `fsDeleteFile()` to `fsRemoveFile()` we change the behavior of MinIO. Previously all the ancestor directories of the removed object were removed if empty up to the root of the bucket.

After the change, the ancestor directories are left untouched. This means after a number of add object - remove object operations we can be left with a bunch of empty directories.

@yauhen-matsiushonak is this an acceptable change in behavior?

Comment (dmitrizpanasas) [reply cmd/panfs.go]: @lech-lorens-epam , sorry for that - I pushed wrong change by mistake. Updated with the right one - please take look again

File: cmd/panfs.go [hunk 2]
```diff
@@ -775,8 +775,19 @@ func (fs *PANFSObjects) DeleteBucket(ctx context.Context, bucket string, opts De
...
-	// Delete all bucket metadata.
-	deleteBucketMetadata(ctx, fs, bucket)
+	if fs.configAgent != nil {
+		noLockID := ""
+		if err = fs.configAgent.DeleteObject(
+			pathJoin(panfsBucketListPrefix, bucket), noLockID); err != nil {
+			return toObjectErr(err, bucket)
+		}
+	} else {
+		if err = fsRemoveAll(ctx, pathJoin(fs.fsPath, minioMetaBucket, bucketMetaPrefix, bucket)); err != nil {
```
Comment (lech-lorens-epam) [inline cmd/panfs.go:781]: The problem you are solving here by adding the `fsRemoveAll()` call is not related to the fact that the config agent is used. If that's OK with you, let's leave the `if fs.configAgent != nil {` part where it originally was and just make sure `fsRemoveAll()` is called before `deleteBucketMetadata()`.

**Review by lech-lorens-epam — CHANGES_REQUESTED** [review]

File: cmd/panfs.go (hunk 2 above)
Comment (dmitrizpanasas) [reply cmd/panfs.go:781]: Does it makes specific sense to have config agent block at the end of the function here?

My thoughts was that it is enough to call `pathJoin(panfsBucketListPrefix, bucket), noLockID)` when config agent is active for delete bucket metadata, that's not right?
